docs/source/notes/autograd.rst

```diff
@@ -3,119 +3,394 @@
 Autograd mechanics
 ==================
 
-This note will present an overview of how autograd works and records the
-operations. It's not strictly necessary to understand all this, but we recommend
-getting familiar with it, as it will help you write more efficient, cleaner
-programs, and can aid you in debugging.
+This note will present an overview of how autograd works. It's not strictly necessary
+to understand all this, but we recommend getting familiar with it, as it will help you
+write more efficient, cleaner programs, and can aid you in debugging.
 
-.. _excluding-subgraphs:
+TL-DR:
 
-Excluding subgraphs from backward
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+- Mathematical formulation: we compute vector Jacobian products using the chain rule. We support inplace operations using hidden versioning of the Tensors.
+- Non-gradients: we provide helper functions to be able to compute things that are not real gradients.
+- Will inplace operation work: Because they modify the original value, if anything needs this value for the backward, the backward will fail.
+- Gradients of 0 vs independent: We don't see the difference.
+- Gradients of 0: They can be represented by a Tensor full of ``0`` s, ``None`` in python, undefined tensor in c++ or and error for ``autograd.grad(allow_unused=False)``.
+- Non-differentiable functions: We have many of those. No guarantee but we try to have a good behavior. Fixed rules on what each elementary operation's backward should compute.
+- Implementation details: Only if you want to understand the current implementation at a high level.
+
+
+
+Mathematical formulation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+What does the autograd package do
+"""""""""""""""""""""""""""""""""
+
+The autograd package allows to use automatic differentiation techniques on top of PyTorch Tensors.
+Automatic differentiation allows to compute dot product between a given vector and the Jacobian of a user-defined function.
+In this document, we will consider the function :math:`f: x \rightarrow y` where :math:`x` is a vector of size :math:`I` and :math:`y` is a vector of size :math:`O`.
+For simplicity, in this document, we will always consider inputs and ouputs to be 1D, in practice, they can be Tensors with an arbitrary number of dimensions.
+The jacobian matrix associated with this function is that matrix :math:`J_f` of size :math:`O \text{x} I` such that each entry is
+given by :math:`(J_f)_{ij} = \dfrac{\partial{x_j}}{\partial{y_i}}`. For simplicity, we will write :math:`J_f = \dfrac{\partial{y}}{\partial{x}}`.
+
+Currently, PyTorch only implements reverse mode automatic differentiation which, given
+an arbitrary vector :math:`v` of size :math:`O`, will compute :math:`v^T J_f`.
```

Comment:
nit: :math:`v^T J_f` is standard standard notation, so no need to change anything, but I prefer other notations: :math:`v \cdot J_f` or :math:`v^i (J_f)_{ij}` with Einstein summation convention or :math:`(v, J_f)`

Reply:
Not sure which one to use. The 60min blitz tutorial [here](https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html#sphx-glr-beginner-blitz-autograd-tutorial-py) uses a mix: :math:`v^T \cdot J_f`
I like explicitly stating the transpose instead of having it implied in the `\cdot`.
But I don't have a very strong opinion here.